### DISCUSSION TITLE
[RFC] Font fallbacks in NextJS

### DISCUSSION THREAD

**Discussion by janicklas-ralph** [vercel/next.js#40112] [Ideas]
# Summary
Build a solution to prevent font based CLS when using `display=swap` when loading Google Fonts.
We can prevent CLS by adjusting the size of the fallback font to match that of the main font using font override metrics such as `size-adjust`, `ascent-override`, `decent-override`, and `line-gap-override` in the `@font-face` rule. You can see a working implementation of this technique in this [demo](https://cultured-rightful-check.glitch.me/).

We are extending the current [fontOptimization](https://nextjs.org/docs/basic-features/font-optimization) in NextJS to add another option experimentalAdjustFallbacks. When enabled, a local fallback font-face definition is generated for each Google font with `display=swap`. The font-face rules will contain the best font override values calculated to adjust the fallback font’s size to closely match the main font.

### Without adjusting font size
![without-override](https://user-images.githubusercontent.com/6142074/187543896-b48d8807-f3c7-4355-b65e-22cb6341a831.gif)

### After adjusting font size
![with-overrides](https://user-images.githubusercontent.com/6142074/187544051-5698c65f-5f9f-411f-8682-c96c7adee740.gif)

# Goal
Sites that load fonts with `font-display: swap` usually suffer from a layout shift ([CLS](https://web.dev/cls/))  when the web font loads and is swapped with the fallback font. This is due to the difference in sizes between the main and fallback fonts. [Malte](https://twitter.com/cramforce) has written an excellent [blog post](http://industrialempathy.com/posts/high-performance-web-font-loading/) outlining the general best practices for loading fonts and describing this [problem while giving us a potential solution](https://www.industrialempathy.com/posts/high-performance-web-font-loading/#what-if-i-really-don't-want-to-use-font-display%3A-optional).

TLDR: we can prevent CLS by adjusting the size of the fallback font to match that of the main font using font override metrics such as `size-adjust`, `ascent-override`, `decent-override`, and `line-gap-override`.

We want to bring this feature (adjusting fallback font size) into NextJS by extending the current Next.js’ [font optimization](https://nextjs.org/docs/basic-features/font-optimization). When this feature is enabled, Next will automatically generate the fallback font-face CSS and inline it during build time.

_Note: As a first step we only handle Google fonts optimization. We will continue to extend this solution to support other fonts in the future._

# Background
`font-display: swap` is generally used to prevent FOIT (Flash of invisible text) and display contents faster on the screen. Swap tells the browser that text using the font should be displayed immediately using a system font, and only when the custom font is ready; replace the system font.

The biggest issue with swap is the jarring effect where the difference in character sizes of the two fonts result in the screen content shifting around. This leads to poor CLS scores of websites with sufficient text content.

We’ve seen that approximately 18% of NextJS sites that use `font-display: swap` also have a poor CLS score. Using `font-display: swap` is one of the primary reasons for an increase in CLS.

# Proposal
The feature detects if your site uses a Google font with `display=swap`, then automatically generates a local fallback font definition whose size will closely match the main font.

This feature is an extension to the current [Font optimization](https://github.com/vercel/next.js/blob/canary/packages/next/server/font-utils.ts) in NextJS, enabled by adding a flag experimentalAdjustFallbacks

To enable in `next.config.js`
```
fontOptimization: {
   // Will generate fallback definition for all Google fonts with display=swap
   experimentalAdjustFallbacks: true
},
```

### Given a config
```
fontOptimization: {
 experimentalAdjustFallbacks: true
},
```

### Fallback font definition is generated automatically with correct size overrides
The font-family name will be generated in the format `{hyphenatedFontName}-fallback`.
Eg: for the font Inter the fallback font will be - inter-fallback, crimson-pro-fallback

_NOTE: The exact CSS that will be generated may change._

```
@font-face {
 font-family: poppins-fallback;
 src: local(Arial);
 ascent-override: 105%;
 descent-override: 35%;
 line-gap-override: 10%;
}
```

### You can use the fallback font in your stylesheets (hyphenated font name + "-fallback")
``` 
body {
 font-family: "Poppins", poppins-fallback, sans-serif;
}
```

## Fallback fonts
Initially we plan to support a single fallback font that works cross platform.
We will start with `Arial` (and `Times New Roman` as the serif fallback) as the local fallback font and plan to extend support to other fallback fonts in the next versions.

# Final output:
The output will look similar to the [Font optimization](https://nextjs.org/docs/basic-features/font-optimization) output, with the fallback override definition included

```
// Injected into html during build/render
<link rel="preconnect" href="https://fonts.gstatic.com" crossorigin />
<style data-href="https://fonts.googleapis.com/css2?family=Inter&display=swap">
 /* latin */
@font-face {
  font-family: 'Inter';
  font-style: normal;
  font-weight: 400;
  font-display: swap;
  src: url(https://fonts.gstatic.com/s/inter/v12/UcCO3FwrK3iLTeHuS_fvQtMwCp50KnMw2boKoduKmMEVuLyfAZ9hiJ-Ek-_EeA.woff2) format('woff2');
  unicode-range: U+0000-00FF, U+0131, U+0152-0153, U+02BB-02BC, U+02C6, U+02DA, U+02DC, U+2000-206F, U+2074, U+20AC, U+2122, U+2191, U+2193, U+2212, U+2215, U+FEFF, U+FFFD;
}
 
 @font-face {
   font-family: inter-fallback,
   src: local('Arial');
   ascent-override: 96.875%;
   descent-override: 24.147%;
   line-gap-override: 0%;
 }
</style>
```

You can now use this fallback font in your CSS as 
```
// Add the fallback font into your font-family definition
body {
 font-family: 'Inter', inter-fallback, sans-serif;
}
```

**Comment by agusterodin** [comment]
Was this behavior present in the past but recently removed? When switching from the old `@next/font` package to the newly built-in `next/font` implementation, the visual similarity of my fallback font is substantially worse.

I am using Montserrat
```
const montserrat = Montserrat({ display: 'swap', weight: '700', fallback: ['Helvetica Neue', 'Helvetica', 'sans-serif'], subsets: ['latin'] })
```

***

Legacy `@next/font/montserrat` (nice, not jarring)

https://github.com/vercel/next.js/assets/10248395/7b5fc998-a683-43eb-8a17-4ffa8ebda2b9

New `next/font/montserrat` (bad, jarring and not visually similar in size)

https://github.com/vercel/next.js/assets/10248395/c0905086-0c44-44ae-abe7-1dd6bfd0d995
